### PR TITLE
State pattern base functions

### PR DESCRIPTION
**Pull request by roman-koshchei**
Made base functions, and children use just different parameters.

### CODE REVIEW

**Review by roman-koshchei — COMMENTED** [review]
all good

File: Client/Pages/UserPage.xaml [hunk 1]
```diff
@@ -185,7 +184,7 @@
                   <RowDefinition />
                 </Grid.RowDefinitions>
 
-                <TextBlock Text="{Binding Text}" Padding="10" Grid.Row="0" TextWrapping="Wrap" />
+                <TextBlock Text="{Binding Text}" Padding="10" Grid.Row="0" Foreground="Black" TextWrapping="Wrap" />
```
Comment (LitvinenkoVitalii) [inline Client/Pages/UserPage.xaml:187]: Not black by default?

File: Client/Services/CommunicationService.cs [hunk 2]
```diff
@@ -67,11 +58,15 @@ public CommunicationService()
         public Action ConfirmSign { get; set; }
 
         // ObservableCollections must not be recreated
-        public ObservableCollection<Prop> Contacts { get; } = new();
+
+        private ObservableCollection<Prop> contacts = new();
```
Comment (LitvinenkoVitalii) [inline Client/Services/CommunicationService.cs:62]: Agreed

File: Client/Services/CommunicationService.cs [hunk 3]
```diff
@@ -67,11 +58,15 @@ public CommunicationService()
...
+        public ObservableCollection<Prop> Contacts { get => contacts; }
 
         public ObservableCollection<Message> CurrentMessages { get; } = new();
         public Action<string> DenySign { get; set; }
-        public ObservableCollection<Prop> Groups { get; } = new();
+
+        private ObservableCollection<Prop> groups = new();
```
Comment (LitvinenkoVitalii) [inline Client/Services/CommunicationService.cs:68]: Agreed

File: Client/Services/CommunicationState.cs [hunk 4]
```diff
@@ -17,6 +15,83 @@ public void SetCommunicationService(CommunicationService cs)
...
+
+                // send data size
+                Request req = new() { Command = Command.GetFileMessageSize, Data = new { To = cs.currentProp.Id, Size = fileReqData.LongLength } };
+                cs.SendData(req);
+
+                cs.pendingSendFile = fileReqData;
+            }
+
+            // abstracts (will be overrided)
+
```
Comment (LitvinenkoVitalii) [inline Client/Services/CommunicationState.cs:94]: Perfect

File: Client/Services/ContactState.cs [hunk 5]
```diff
@@ -1,11 +1,4 @@
-﻿using Client.Models;
-using CrossLibrary;
-using Newtonsoft.Json;
-using System.Collections.Generic;
-using System.IO;
-using static CrossLibrary.Globals;
-
-namespace Client.Services
+﻿namespace Client.Services
```
Comment (LitvinenkoVitalii) [inline Client/Services/ContactState.cs:1]: lost using

File: Client/Services/CommunicationService.cs [hunk 6]
```diff
@@ -1,13 +1,4 @@
...
-using Client.Network;
-using CrossLibrary;
-using Newtonsoft.Json;
-using System;
-using System.Collections.Generic;
-using System.Collections.ObjectModel;
-using System.Windows;
-
-namespace Client.Services
+﻿namespace Client.Services
```
Comment (LitvinenkoVitalii) [inline Client/Services/CommunicationService.cs:1]: lost using

File: Server/Services/CommunicationService.cs [hunk 7]
```diff
@@ -1,19 +1,11 @@
-using CrossLibrary;
```
Comment (LitvinenkoVitalii) [inline Server/Services/CommunicationService.cs:1]: lost using

File: Client/Services/GroupState.cs [hunk 8]
```diff
@@ -1,11 +1,4 @@
-﻿using Client.Models;
-using CrossLibrary;
-using Newtonsoft.Json;
-using System.Collections.Generic;
-using System.IO;
-using static CrossLibrary.Globals;
-
-namespace Client.Services
+﻿namespace Client.Services
```
Comment (LitvinenkoVitalii) [inline Client/Services/GroupState.cs:1]: lost using

File: Client/Services/GroupState.cs [hunk 9]
```diff
@@ -17,69 +10,22 @@ public override void Leave(int id)
...
-
-                for (int i = 0; i < cs.Groups.Count; i++)
-                {
-                    if (cs.Groups[i].Id == cs.currentProp.Id)
-                    {
-                        cs.Groups[i] = cs.currentProp;
-                        break;
-                    }
-                }
+                var data = new { Id = cs.currentProp.Id, NewName = newName };
```
Comment (LitvinenkoVitalii) [inline Client/Services/GroupState.cs:17]: var or dynamic?

File: Client/Services/ContactState.cs [hunk 10]
```diff
@@ -17,70 +10,22 @@ public override void Leave(int id)
...
-                for (int i = 0; i < cs.Contacts.Count; i++)
-                {
-                    if (cs.Contacts[i].Id == cs.currentProp.Id)
-                    {
-                        cs.currentProp = new Prop { Id = cs.currentProp.Id, Name = newName };
-                        cs.Contacts[i] = cs.currentProp;
-                        break;
-                    }
-                }
+                var data = new { To = cs.currentProp.Id, From = cs.User.Id, NewName = newName };
```
Comment (LitvinenkoVitalii) [inline Client/Services/ContactState.cs:17]: var or dynamic?

File: Client/Services/GroupState.cs [hunk 11]
```diff
@@ -17,69 +10,22 @@ public override void Leave(int id)
...
 
-            public override void RefreshMessages()
-            {
-                cs.CurrentMessages.Clear();
-                foreach (Message message in cs.groupMessages[cs.currentProp.Id])
-                {
-                    cs.CurrentMessages.Add(message);
-                }
-            }
+            public override void RefreshMessages() => RefreshMessages(ref cs.groupMessages);
```
Comment (LitvinenkoVitalii) [inline Client/Services/GroupState.cs:13]: Agreed

File: Client/Services/GroupState.cs [hunk 12]
```diff
@@ -17,69 +10,22 @@ public override void Leave(int id)
...
 
             public override void SendMessage(string messageStr)
             {
-                Message message = new() { Text = messageStr, IsIncoming = false };
-                cs.groupMessages[cs.currentProp.Id].AddLast(message);
-                cs.CurrentMessages.Add(message);
-
-                Request req = new() { Command = Command.SendMessageToGroup, Data = new { To = cs.currentProp.Id, From = cs.User.Id, Message = message.Text } };
-                cs.SendData(req);
+                SendMessage(messageStr, ref cs.groupMessages, Command.SendMessageToGroup);
```
Comment (LitvinenkoVitalii) [inline Client/Services/GroupState.cs:28]: Agreed

File: Client/Services/GroupState.cs [hunk 13]
```diff
@@ -17,69 +10,22 @@ public override void Leave(int id)
...
-                Request fileReq = new() { Command = Command.SendFileMessageToGroup, Data = new { From = cs.User.Id, Message = message.Text, Files = files } };
-                string fileReqStr = JsonConvert.SerializeObject(fileReq);
-                byte[] fileReqData = CommunicationEncoding.GetBytes(fileReqStr);
-
-                // send data size
-                Request req = new() { Command = Command.GetFileMessageSize, Data = new { To = cs.currentProp.Id, Size = fileReqData.LongLength } };
-                cs.SendData(req);
-
-                cs.pendingSendFile = fileReqData;
+                SendFileMessage(messageStr, filePaths, ref cs.groupMessages, Command.SendFileMessageToGroup);
```
Comment (LitvinenkoVitalii) [inline Client/Services/GroupState.cs:23]: Perfect

File: Client/Services/ContactState.cs [hunk 14]
```diff
@@ -17,70 +10,22 @@ public override void Leave(int id)
...
-                Request fileReq = new() { Command = Command.SendFileMessageToContact, Data = new { From = cs.User.Id, Message = message.Text, Files = files } };
-                string fileReqStr = JsonConvert.SerializeObject(fileReq);
-                byte[] fileReqData = CommunicationEncoding.GetBytes(fileReqStr);
-
-                // send data size
-                Request req = new() { Command = Command.GetFileMessageSize, Data = new { To = cs.currentProp.Id, Size = fileReqData.LongLength } };
-                cs.SendData(req);
-
-                cs.pendingSendFile = fileReqData;
+                SendFileMessage(messageStr, filePaths, ref cs.contactMessages, Command.SendFileMessageToContact);
```
Comment (LitvinenkoVitalii) [inline Client/Services/ContactState.cs:23]: Agreed

File: Client/Services/CommunicationService.cs [hunk 15]
```diff
@@ -16,10 +7,10 @@ public partial class CommunicationService
         private CommunicationState state;
 
         // key = contact id
-        private readonly Dictionary<int, LinkedList<Message>> contactMessages = new();
+        private Dictionary<int, LinkedList<Message>> contactMessages = new();
```
Comment (LitvinenkoVitalii) [inline Client/Services/CommunicationService.cs:10]: Agreed

File: Client/Services/CommunicationService.cs [hunk 16]
```diff
@@ -16,10 +7,10 @@ public partial class CommunicationService
...
 
         // key = group id
-        private readonly Dictionary<int, LinkedList<Message>> groupMessages = new();
+        private Dictionary<int, LinkedList<Message>> groupMessages = new();
```
Comment (LitvinenkoVitalii) [inline Client/Services/CommunicationService.cs:13]: Agreed

**Review by LitvinenkoVitalii — APPROVED** [review]

**Pull request merged**